### PR TITLE
add __cxa_init_primary_exception

### PR DESCRIPTION
**Pull request by itrofimow**
Aims to resolve https://github.com/libcxxrt/libcxxrt/issues/22

Somewhat related libc++ patch: https://reviews.llvm.org/D158620

### CODE REVIEW

**Comment by davidchisnall** [comment]
That looks plausible. Can you add a few tests? I’d like to see, at least, that this exception can be thrown and caught and that the destructor runs in the right place.

File: test/test_exception.cc [hunk 1]
```diff
@@ -322,6 +322,103 @@ void test_uncaught_exceptions()
...
+	template <typename Ex>
+	static void Dest(void *e) {
+		static_cast<Ex*>(e)->~Ex();
+	}
+
+	void *ex_ = nullptr;
+};
+
+template <typename Ex>
+ExceptionPtr MakeExceptionPtr(Ex e) {
```
Comment (itrofimow) [inline test/test_exception.cc:366]: basically what gcc does: https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/libsupc%2B%2B/exception_ptr.h#L253

**Comment by itrofimow** [comment]
@davidchisnall added a `std::exception_ptr`-alike test.
Could you please have a look?

**Review by davidchisnall — APPROVED** [review]
Looks great, thank you.

@emaste , do you think we could sneak this into FreeBSD 14.0? We can do the libc++ updates a bit later, but it would be nice to have the new APIs in so that external things can depend on them, even if the base system libc++ doesn’t yet.

**Comment by emaste** [comment]
> @emaste , do you think we could sneak this into FreeBSD 14.0?

I think there's still time, we can sync this repo into FreeBSD main and MFC from there.

**Comment by davidchisnall** [comment]
[ Rebased to pick up the fix to make FreeBSD CI work ]

**Comment by itrofimow** [comment]
Hi! Is there anything holding this PR back from being merged?

I would like to try to make use of these changes in https://github.com/llvm/llvm-project/pull/65534, and it would be a bit more convenient to do so with master branch of libcxxrt rather than with a fork

**Comment by davidchisnall** [comment]
Sorry, I thought I set it to merge when the CI finished.

**Pull request merged**